internal/emlibc/resolver.go

```diff
@@ -0,0 +1,114 @@
+package emlibc
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/go-interpreter/wagon/exec"
+
+	"github.com/go-interpreter/wagon/wasm"
+)
+
+func ResolveEnv(name string) (*wasm.Module, error) {
+	if name == "env" {
+		return GetEnv(), nil
+	}
+	fmt.Println("tried resolve", name)
+	return nil, errors.New("Not Found")
+}
+func clen(n []byte) int {
+	for i := 0; i < len(n); i++ {
+		if n[i] == 0 {
+			return i
+		}
+	}
+	return len(n)
+}
+func GetEnv() *wasm.Module {
+
+	m := wasm.NewModule()
+	print := func(proc *exec.Process, v int32) int32 {
+		fmt.Printf("result = %v\n", v)
+		return 0
+	}
+	puts := func(proc *exec.Process, v int32) int32 {
+
+		buf := []byte{}
+		temp := make([]byte, 1)
```

Comment:
Maybe try `var temp [1]byte` and reference it like `temp[:]` ?

Reply:
yeah i like that...I couldn't decide if it would be better if we had more access to the underlaying []byte, I guess at some point during this process maybe we need to implement a couple other methods on exec.Process for memory management, that may be where we can implement things like alloc/free etc. and could also have a method that returns a reader so we could use bufio.Readers for some of this

Reply:
> that may be where we can implement things like alloc/free etc

Hmmmm. How do things like rust/Go handle this? Does everyone who compiles wasm ship their own malloc/free implementation?

Reply:
Thats been one of the most confusing things about this as I have been learning WASM. From what I can gather it currently ships inside the glue code from compiling. So its actually provided by the compiler 'runtime'...like emscripten or LLVM. 
Here is some of the code out of the emscripten js file that provides the environment to run the wasm
```js
var _free = Module["_free"] = function() {
  return Module["asm"]["_free"].apply(null, arguments)
};

var _main = Module["_main"] = function() {
  return Module["asm"]["_main"].apply(null, arguments)
};

var _malloc = Module["_malloc"] = function() {
  return Module["asm"]["_malloc"].apply(null, arguments)
};

var _memcpy = Module["_memcpy"] = function() {
  return Module["asm"]["_memcpy"].apply(null, arguments)
};

var _memset = Module["_memset"] = function() {
  return Module["asm"]["_memset"].apply(null, arguments)
};
```
and then its called from the wasm
```wat
    call $_printf
    i32.const 4
    call $_malloc
    local.set 4
    local.get 1
    local.get 4
    i32.store
    local.get 1
```
Really seems like it should have been part of the MVP spec to me...that seems pretty basic, but it appears to me that is how its done...

But its an area of WASM i'm still trying to learn.

Reply:
I've learned the most about wasm from playing around with https://github.com/intel/wasm-micro-runtime
They support building with emscripten and clang. Its a pretty clean implementation but targeted more at resource constrained environments.

But to support emscripten (and llvm) they have a libc wrapper and it is the most concise place i've found for figuring out what calls are needed to support the compiler runtimes.

In the following code you can see their "env" implementation. 

https://github.com/intel/wasm-micro-runtime/blob/307b54cc5946a5d07ee17bc177c1a2f17e231836/core/iwasm/lib/native/libc/libc_wrapper.c#L933-L969

It covers the basics for libc calls from emscripten or llvm...and from most of the code i've tried against it works without issue. You can clearly see malloc and free operating inside the memory buffer. So thats what i'm basing my info off of.

Reply:
FWICT, wasm is intended just to implement a very minimal 'CPU', and design decisions like memory allocation are to be handled by the calling code. The only memory management features in the wasm specification are these two opcodes:

 - memory.grow - increase the size of the memory buffer.
 - memory.current - return the current size of the memory buffer.

In all the wasm I've seen, malloc/free are all implemented in wasm shipped by the application.




Do any other wasm interpreters provide a 'libc' layer like this, or is the libc layer always shipped with the application code?

Reply:
I agree with what you say from the spec

yes wasmer appears to provide a 'emscripten compatibility layer' and wasm-micro-runtime does as well.

I think it is kind of a grey area at the moment but the problem I see it that since the linear memory buffer is used from both the host and the wasm module then someone has to be authoritative

i.e. if I call into a wasm module from the host and want to pass in a string, I allocate it in the buffer and send a pointer.

If they reply with another string they basically do the same. 

So do we assume the linear memory is stateless? So in each iteration the current 'owner' has full use of the buffer...if not someone has to manage the memory. I think that since GC is planned in the post MVP then I think the responsibility for memory management is best handled in the runtime host.

Again these are my very very unqualified opinions

Reply:
One thing to point out, if you build for a browser then the libc code is in the .js glue file generated by emscripten, so the browser does not directly provide it.

but in the case of non browser runtimes it appears to me they have provided their own glue code natively to support the emscripten compiler.

that is why i mentioned doing the implementations as a library of .wasm files that could be imported rather than writing them natively in go...but we would need to provide at least a minimal api to do that

Reply:
Finally found the wasmer code...I remembered seeing it but it took me a minute

https://github.com/wasmerio/wasmer/blob/56c571465ea65992172897da44ad2970e0ec55a0/lib/emscripten/src/lib.rs#L555

Reply:
> but in the case of non browser runtimes it appears to me they have provided their own glue code natively to support the emscripten compiler.
> 
> that is why i mentioned doing the implementations as a library of .wasm files that could be imported rather than writing them natively in go...but we would need to provide at least a minimal api to do that

Lets do that - (as in, lets do what wasmer and the other native runtimes are doing, and provide an identical API).